web/source/osk/visualKeyboard.ts

```diff
@@ -0,0 +1,2669 @@
+        btn.onmouseup=btn.onmouseout=osk.mouseUpMouseOutHandler; //Build 360
+      }
+
+      // Handle subkey-related tasks.
+      if(typeof(spec['sk']) != 'undefined' && spec['sk'] != null) {
+        this.processSubkeys(btn);
+      } else {
+        btn['subKeys']=null;
+      }
+      
+      // Add text to button and button to placeholder div
+      btn.appendChild(this.generateKeyText(osk));
+      kDiv.appendChild(btn);
+
+      // Prevent user selection of key captions
+      //t.style.webkitUserSelect='none';
+      
+      // The 'return value' of this process.
+      return {element: kDiv, percent: totalPercent - originalPercent};
+    }
+
+    objectGeometry(v: number): string {
+      let unit = this.objectUnits();
+      if(unit == '%') {
+        return v + unit;
+      } else { // unit == 'px'
+        return Math.round(v)+unit;
+      }
+    }
+  }
+
+  export class OSKSubKey extends OSKKey {
+    constructor(spec: OSKKeySpec) {
+      super(spec);
+    }
+
+    getId(osk: VisualKeyboard): string {
+      let spec = this.spec;
+      // Create (temporarily) unique ID by prefixing 'popup-' to actual key ID
+      if(typeof(spec['layer']) == 'string' && spec['layer'] != '') {
+        return 'popup-'+spec['layer']+'-'+spec['id'];
+      } else {
+        // We only create subkeys when they're needed - the currently-active layer should be fine.
+        return 'popup-' + osk.layerId + '-'+spec['id'];
+      }
+    }
+
+    construct(osk: VisualKeyboard, baseKey: HTMLDivElement, topMargin: boolean): HTMLDivElement {
+      let spec = this.spec;
+
+      let kDiv=document.createElement('div');
+      let tKey = osk.getDefaultKeyObject();
+      let ks=kDiv.style;
+
+      for(var tp in tKey) {
+        if(typeof spec[tp] != 'string') {
+          spec[tp]=tKey[tp];
+        }
+      }
+
+      kDiv.className='kmw-key-square-ex';
+      kDiv['keyId']=spec['id'];
+      if(topMargin) {
+        ks.marginTop='5px';
+      }
+
+      if(typeof spec['width'] != 'undefined') {
+        ks.width=(parseInt(spec['width'],10)*baseKey.offsetWidth/100)+'px';
+      } else {
+        ks.width=baseKey.offsetWidth+'px';
+      }
+      ks.height=baseKey.offsetHeight+'px';
+
+      let btn=document.createElement('div');
+      osk.setButtonClass(spec,btn);
+
+      btn.id = this.getId(osk);
+      btn['key'] = this;
+
+      // Must set button size (in px) dynamically, not from CSS
+      let bs=btn.style;
+      bs.height=ks.height;
+      bs.width=ks.width;
+
+      // Must set position explicitly, at least for Android
+      bs.position='absolute';
+
+      btn.appendChild(this.generateKeyText(osk));
+      kDiv.appendChild(btn);
+
+      return kDiv;
+    }
+  }
+
+  //#endregion
+
+  export class VisualKeyboard {
+    //#region Keyboard-related static constants
+    // Define Keyman Developer modifier bit-flags (exposed for use by other modules)
+    static modifierCodes = {
+      "LCTRL":0x0001,
```

Comment:
KMEA references this mapping. Can you update the comment in
/android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMModifierCodes.java line 7?

Reply:
I'll address that in https://github.com/keymanapp/keyman/pull/1525, since it will re-relocate the field.